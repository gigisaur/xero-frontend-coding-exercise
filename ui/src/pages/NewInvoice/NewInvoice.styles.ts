import { Form } from "formik";
import styled from "styled-components";
import { ButtonBase, Color, FontFamily } from "../../theme";

export const StyledForm = styled(Form)`
    display: grid;
    align-items: end;
    grid-template-columns: repeat(auto-fit, minmax(12.5rem, 1fr));
    grid-row-gap: 0.5rem;
    grid-column-gap: 0.5rem;
    margin-bottom: 2rem;
`;

export const Main = styled.main`
    background-color: ${Color.PageBackground};
    border-top: 0.5rem solid ${Color.Accent};
    width: 100%;
    font-family: ${FontFamily};
`;

export const Article = styled.article`
    display: flex;
    flex-direction: column;
    margin: auto;
    max-width: 50%;
    min-width: 18.75rem;
    min-height: 31.25rem;
    margin-top: 5%;
    background-color: ${Color.ContentBackground};
    box-shadow: 0 -2px 0.75rem rgba(0, 0, 0, 0.03);
    padding: 1rem;
`;

export const TotalCostP = styled.p`
    align-items: end;
    margin-top: 2rem;
    align-content: end;
    display: grid;
    flex: 1;
    justify-content: end;
`;

export const SubmitInvoiceButton = styled(ButtonBase)`
    color: white;
    background-color: ${Color.Accent};

    &:hover {
        background-color: ${Color.PrimaryButtonHoverBackground};
    }
`;

export const AddItemButton = styled(ButtonBase)`
    color: ${Color.Accent};
    background-color: white;
    border: 1px solid ${Color.Accent};
    padding: 0 0.9rem;
    align-self: start;
    margin-top: 1.25rem;

    &:hover {
        background-color: ${Color.Accent};
        color: white;
    }
`;

export const SubmitInvoiceButtonContainerDiv = styled.div`
    display: grid;
    justify-items: end;
`;
