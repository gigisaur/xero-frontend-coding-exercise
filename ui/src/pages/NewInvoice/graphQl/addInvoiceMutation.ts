import gql from "graphql-tag";
import { IInvoice } from "..";

export interface IAddInvoiceResult {
    addInvoice: IInvoice;
}

export interface IAddInvoiceVariables {
    [key: string]: IInvoice;
}

export const addInvoiceMutation = gql`
    mutation AddInvoice($params: InvoiceCreateInput!) {
        addInvoice(params: $params) {
            items {
                description
                cost
                quantity
                price
            }
            totalCost
        }
    }
`;
