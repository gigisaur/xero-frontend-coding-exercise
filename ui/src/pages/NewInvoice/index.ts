export * from "./NewInvoice";

export interface IFormValues {
    description: string;
    cost: number;
    quantity: number;
}

export interface IItem extends IFormValues {
    price: number;
}

export interface IInvoice {
    items: IItem[];
    totalCost: number;
}
