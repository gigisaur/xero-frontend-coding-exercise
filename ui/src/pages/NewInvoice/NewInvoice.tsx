import { useMutation } from "@apollo/react-hooks";
import { Formik, FormikProps } from "formik";
import React, { useState } from "react";
import { useToasts } from "react-toast-notifications";
import * as Yup from "yup";
import { IFormValues, IInvoice, IItem } from ".";
import { InputFormik } from "../../components/InputFormik";
import { Table, Td, Th, Tr } from "../../components/Table";
import { strings } from "../../localization";
import { addInvoiceMutation, IAddInvoiceResult, IAddInvoiceVariables } from "./graphQl/addInvoiceMutation";
import { AddItemButton, Article, Main, StyledForm, SubmitInvoiceButton , SubmitInvoiceButtonContainerDiv, TotalCostP } from "./NewInvoice.styles";

export const NewInvoice = () => {
    const { addToast } = useToasts();
    const [state, setState] = useState<IInvoice>({ items: [], totalCost: 0 });
    const [submitInvoice] = useMutation<IAddInvoiceResult, IAddInvoiceVariables>(addInvoiceMutation);

    const validationSchemaObj = Yup.object().shape({
        description: Yup.string().required(strings.Message_Description_RequiredValidation),
        cost: Yup.number().min(0, strings.Message_Cost_PositiveValidation).required(strings.Message_Cost_RequiredValidation),
        quantity: Yup.number().required(strings.Message_Quantity_RequiredValidation),
    });

    const initialValues: IFormValues = {
        description: "",
        cost: 0,
        quantity: 0,
    }

    const handleOnSubmit = (values: IFormValues) => {
        const newItem: IItem = {
            ...values,
            price: values.cost * values.quantity,
        };
        setState({ items: state.items.concat(newItem), totalCost: state.totalCost + newItem.price });
    }

    const handleOnSubmitInvoiceClick = () => {
        if (!state.items || state.items.length === 0) return;

        submitInvoice({ variables: { params: state }})
            .then(() => {
                addToast(strings.Message_InvoiceSubmitted, { appearance: "success", autoDismiss: true });
                setState({ items: [], totalCost: 0 });
            })
            .catch(() => addToast(strings.Message_UnknownError, { appearance: "error", autoDismiss: true }));
    }

    return (
        <Main>
            <Article>
                <Formik
                    initialValues={initialValues}
                    onSubmit={handleOnSubmit}
                    validateOnBlur
                    validationSchema={validationSchemaObj}
                >
                    {(formikProps: FormikProps<any>) => (
                        <StyledForm>
                            <InputFormik
                                type="text"
                                label={strings.Label_Description}
                                formikProps={formikProps}
                                name="description"
                                required
                            />
                            <InputFormik
                                type="number"
                                label={strings.Label_Cost}
                                formikProps={formikProps}
                                name="cost"
                                required
                            />
                            <InputFormik
                                type="number"
                                label={strings.Label_Quantity}
                                formikProps={formikProps}
                                name="quantity"
                                required
                            />
                            <AddItemButton type="submit">{strings.Action_AddItem}</AddItemButton>
                        </StyledForm>
                    )}
                </Formik>

                <Table>
                    <thead>
                        <tr>
                            <Th>{strings.Label_Description}</Th>
                            <Th>{strings.Label_Cost}</Th>
                            <Th>{strings.Label_Quantity}</Th>
                            <Th>{strings.Label_Price}</Th>
                        </tr>
                    </thead>
                    <tbody>
                        {state.items.map(({ cost, description, quantity, price }: IItem, index: number) => (
                            // Index as key isn't ideal but it is safe here as we are only adding to the end of the list. To improve add an ID to the items and use it
                            <Tr key={index}>
                                <Td>{description}</Td>
                                <Td>${cost}</Td>
                                <Td>{quantity}</Td>
                                <Td>${price}</Td>
                            </Tr>
                        ))}
                    </tbody>
                </Table>

                <TotalCostP>{strings.formatString(strings.Label_TotalCost, state.totalCost)}</TotalCostP>

                <SubmitInvoiceButtonContainerDiv>
                    <SubmitInvoiceButton
                        type="button"
                        onClick={handleOnSubmitInvoiceClick}
                        aria-disabled={state.items.length === 0}
                    >
                        {strings.Action_SubmitInvoice}
                    </SubmitInvoiceButton>
                </SubmitInvoiceButtonContainerDiv>
            </Article>
        </Main>
    )
};
