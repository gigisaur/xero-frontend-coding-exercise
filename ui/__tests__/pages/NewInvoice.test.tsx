import { MockedProvider } from "@apollo/react-testing";
import { act, fireEvent, render, wait } from "@testing-library/react";
import { axe } from "jest-axe";
import "jest-styled-components";
import * as React from "react";
import { ToastProvider } from "react-toast-notifications";
import { NewInvoice } from "../../src/pages/NewInvoice";
import { addInvoiceMutation } from "../../src/pages/NewInvoice/graphQl/addInvoiceMutation";
import { addInvoiceMutationErrorMocks, addInvoiceMutationMocks, params, result } from "../_data/newInvoiceMutationData";

// Intentionally leaving InputFormik and toasts unmocked for ease of testing

test(`When rendered then the HTML matches this snapshot`, () => {
    const { container, unmount } = render(
        <MockedProvider mocks={addInvoiceMutationMocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );

    expect(container).toMatchSnapshot();
    unmount();
});

test(`Given form fields with values when the Add Item button is clicked then the item is added to the table and the Submit Invoice button is enabled`, async () => {
    const { container, unmount, getByLabelText, getByText } = render(
        <MockedProvider mocks={addInvoiceMutationMocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );

    // Arrange the form
    const descriptionInput = getByLabelText(/Description/g);
    const costInput = getByLabelText(/Cost/g);
    const quantityInput = getByLabelText(/Quantity/g);

    act(() => { fireEvent.change(descriptionInput, { target: { value: "Apples" } }) });
    act(() => { fireEvent.change(costInput, { target: { value: 10 } }) });
    act(() => { fireEvent.change(quantityInput, { target: { value: 2 } }) });

    // Act
    await wait(() => {
        fireEvent.click(getByText("Add item"));
    });

    // Assert
    expect(container).toMatchSnapshot();
    expect(getByText("Submit invoice").getAttribute("aria-disabled")).toEqual("false");

    unmount();
});

test(`Given there are no items in the table then the Submit Invoice button is aria-disabled and clicking it does not fire a mutation`, () => {
    const mutationSpy = jest.fn();

    const mocks = [
        {
            request: {
                query: addInvoiceMutation,
                variables: { params },
            },
            result: () => {
                mutationSpy();
                return { data: { result } };
            },
        },
    ];

    const { getByText, unmount } = render(
        <MockedProvider mocks={mocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );

    // Assert
    expect(mutationSpy).toHaveBeenCalledTimes(0);

    // Act
    act(() => {
        fireEvent.click(getByText("Submit invoice"));
    })

    // Assert
    expect(mutationSpy).toHaveBeenCalledTimes(0);
    expect(getByText("Submit invoice").getAttribute("aria-disabled")).toEqual("true");
    unmount();
});

test(`Given an item with cost $5 and quantity 2 then the price should be $10`, async () => {
    const { unmount, getByLabelText, getByText } = render(
        <MockedProvider mocks={addInvoiceMutationMocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );

    // Arrange the form
    const descriptionInput = getByLabelText(/Description/g);
    const costInput = getByLabelText(/Cost/g);
    const quantityInput = getByLabelText(/Quantity/g);

    act(() => { fireEvent.change(descriptionInput, { target: { value: "Apples" } }) });
    act(() => { fireEvent.change(costInput, { target: { value: 5 } }) });
    act(() => { fireEvent.change(quantityInput, { target: { value: 2 } }) });

    // Act
    await wait(() => {
        fireEvent.click(getByText("Add item"));
    });

    // Assert
    expect(getByText("$10"));

    unmount();
});

test(`Given two items with a combined price of $30 then the Total Cost should read $30`, async () => {
    const { unmount, getByLabelText, getByText } = render(
        <MockedProvider mocks={addInvoiceMutationMocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );

    // Arrange the form
    const descriptionInput = getByLabelText(/Description/g);
    const costInput = getByLabelText(/Cost/g);
    const quantityInput = getByLabelText(/Quantity/g);

    act(() => { fireEvent.change(descriptionInput, { target: { value: "Apples" } }) });
    act(() => { fireEvent.change(costInput, { target: { value: 10 } }) });
    act(() => { fireEvent.change(quantityInput, { target: { value: 2 } }) });

    await wait(() => {
        fireEvent.click(getByText("Add item"));
    });

    act(() => { fireEvent.change(descriptionInput, { target: { value: "Bananas" } }) });
    act(() => { fireEvent.change(costInput, { target: { value: 10 } }) });
    act(() => { fireEvent.change(quantityInput, { target: { value: 1 } }) });

    await wait(() => {
        fireEvent.click(getByText("Add item"));
    });

    // Assert
    expect(getByText("Total: $30"));

    unmount();
});

test(`Given an item has been added when the Submit Invoice button is clicked then the submitInvoice mutation is fired`, async () => {
    const mutationSpy = jest.fn();

    const mocks = [
        {
            request: {
                query: addInvoiceMutation,
                variables: { params },
            },
            result: () => {
                mutationSpy();
                return { data: result };
            },
        },
    ];

    const { getByText, getByLabelText, unmount } = render(
        <MockedProvider mocks={mocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );

    // Arrange the form
    const descriptionInput = getByLabelText(/Description/g);
    const costInput = getByLabelText(/Cost/g);
    const quantityInput = getByLabelText(/Quantity/g);

    act(() => { fireEvent.change(descriptionInput, { target: { value: "Apples" } }) });
    act(() => { fireEvent.change(costInput, { target: { value: 10 } }) });
    act(() => { fireEvent.change(quantityInput, { target: { value: 2 } }) });

    await wait(() => {
        fireEvent.click(getByText("Add item"));
    });

    // Assert
    expect(mutationSpy).toHaveBeenCalledTimes(0);

    // Act
    await wait(() => {
        fireEvent.click(getByText("Submit invoice"));
    })

    await wait();

    // Assert
    expect(mutationSpy).toHaveBeenCalledTimes(1);
    expect(getByText("Submit invoice").getAttribute("aria-disabled")).toEqual("true");
    unmount();
});

test(`When the Submit Invoice mutation returns successfully then a confirmation toast is shown`, async () => {
    const { getByText, getByLabelText, unmount } = render(
        <MockedProvider mocks={addInvoiceMutationMocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );

    // Arrange the form
    const descriptionInput = getByLabelText(/Description/g);
    const costInput = getByLabelText(/Cost/g);
    const quantityInput = getByLabelText(/Quantity/g);

    act(() => { fireEvent.change(descriptionInput, { target: { value: "Apples" } }) });
    act(() => { fireEvent.change(costInput, { target: { value: 10 } }) });
    act(() => { fireEvent.change(quantityInput, { target: { value: 2 } }) });

    await wait(() => {
        fireEvent.click(getByText("Add item"));
    });

    // Act
    await wait(() => {
        fireEvent.click(getByText("Submit invoice"));
    })

    await wait();

    // Assert
    expect(getByText("Invoice submitted"));
    unmount();
});

test(`When the Submit Invoice mutation errors then an error toast is shown`, async () => {
    const { getByText, getByLabelText, unmount } = render(
        <MockedProvider mocks={addInvoiceMutationErrorMocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );

    // Arrange the form
    const descriptionInput = getByLabelText(/Description/g);
    const costInput = getByLabelText(/Cost/g);
    const quantityInput = getByLabelText(/Quantity/g);

    act(() => { fireEvent.change(descriptionInput, { target: { value: "Apples" } }) });
    act(() => { fireEvent.change(costInput, { target: { value: 10 } }) });
    act(() => { fireEvent.change(quantityInput, { target: { value: 2 } }) });

    await wait(() => {
        fireEvent.click(getByText("Add item"));
    });

    // Act
    await wait(() => {
        fireEvent.click(getByText("Submit invoice"));
    })

    await wait();

    // Assert
    expect(getByText("Something went wrong"));
    unmount();
});

test(`When rendered then there are no accessibility violations`, async () => {
    const { container, unmount } = render(
        <MockedProvider mocks={addInvoiceMutationMocks} addTypename={false}>
            <ToastProvider>
                <NewInvoice />
            </ToastProvider>
        </MockedProvider>
    );
    const results = await axe(container)

    expect(results).toHaveNoViolations();
    unmount();
});
