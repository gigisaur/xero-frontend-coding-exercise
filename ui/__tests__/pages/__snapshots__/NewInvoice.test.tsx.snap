// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given form fields with values when the Add Item button is clicked then the item is added to the table and the Submit Invoice button is enabled 1`] = `
.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #191919;
  font-size: 0.9rem;
}

.c6 {
  width: 100%;
  border: 1px solid #DAE5EC;
  background-color: white;
  border-radius: 0.25rem;
  line-height: 1.3em;
  font-size: 0.9rem;
  text-indent: 0.8rem;
  padding: 0;
  box-sizing: border-box;
  font-family: Roboto,Arial,Helvetica,sans-serif;
  color: #191919;
  height: 2.5rem;
}

.c7 {
  line-height: 0.9rem;
  padding-bottom: 1rem;
  padding-top: 0.2rem;
  font-size: 0.8rem;
}

.c3 {
  display: grid;
  grid-template-rows: 1.25rem minmax(2.5rem,auto) minmax(2.1rem,auto);
  box-sizing: border-box;
}

.c5 {
  margin-left: 0.25rem;
  color: #E23636;
}

.c9 {
  border-collapse: collapse;
  width: 100%;
  border: 1px solid #EBF6FF;
}

.c12 {
  padding: 0.5rem;
  color: #191919;
  font-size: 0.9rem;
  border-bottom: 1px solid #EBF6FF;
}

.c10 {
  padding: 0.5rem;
  border-bottom: 1px solid #DAE5EC;
  color: #191919;
  opacity: 0.75;
  font-size: 0.8rem;
  text-transform: uppercase;
  text-align: left;
}

.c11:nth-child(odd) {
  background-color: #EBF6FF;
}

.c2 {
  display: grid;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  grid-template-columns: repeat(auto-fit,minmax(12.5rem,1fr));
  grid-row-gap: 0.5rem;
  grid-column-gap: 0.5rem;
  margin-bottom: 2rem;
}

.c0 {
  background-color: #e9ebec;
  border-top: 0.5rem solid #00b7e2;
  width: 100%;
  font-family: Roboto,Arial,Helvetica,sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: auto;
  max-width: 50%;
  min-width: 18.75rem;
  min-height: 31.25rem;
  margin-top: 5%;
  background-color: #fafafa;
  box-shadow: 0 -2px 0.75rem rgba(0,0,0,0.03);
  padding: 1rem;
}

.c13 {
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  margin-top: 2rem;
  -webkit-align-content: end;
  -ms-flex-line-pack: end;
  align-content: end;
  display: grid;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: end;
  -webkit-justify-content: end;
  -ms-flex-pack: end;
  justify-content: end;
}

.c15 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 2.5rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all ease-in-out 0.2s;
  transition: all ease-in-out 0.2s;
  vertical-align: top;
  white-space: nowrap;
  font-weight: bold;
  padding: 0 1.5rem;
  position: relative;
  font-size: 1rem;
  text-transform: capitalize;
  -webkit-letter-spacing: 0.01rem;
  -moz-letter-spacing: 0.01rem;
  -ms-letter-spacing: 0.01rem;
  letter-spacing: 0.01rem;
  font-family: inherit;
  color: white;
  background-color: #00b7e2;
}

.c15[aria-disabled='true'] {
  opacity: 0.4;
  pointer-events: none;
}

.c15:hover {
  background-color: #0097bb;
}

.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 2.5rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all ease-in-out 0.2s;
  transition: all ease-in-out 0.2s;
  vertical-align: top;
  white-space: nowrap;
  font-weight: bold;
  padding: 0 1.5rem;
  position: relative;
  font-size: 1rem;
  text-transform: capitalize;
  -webkit-letter-spacing: 0.01rem;
  -moz-letter-spacing: 0.01rem;
  -ms-letter-spacing: 0.01rem;
  letter-spacing: 0.01rem;
  font-family: inherit;
  color: #00b7e2;
  background-color: white;
  border: 1px solid #00b7e2;
  padding: 0 0.9rem;
  -webkit-align-self: start;
  -ms-flex-item-align: start;
  align-self: start;
  margin-top: 1.25rem;
}

.c8[aria-disabled='true'] {
  opacity: 0.4;
  pointer-events: none;
}

.c8:hover {
  background-color: #00b7e2;
  color: white;
}

.c14 {
  display: grid;
  justify-items: end;
}

<div>
  <main
    class="c0"
  >
    <article
      class="c1"
    >
      <form
        action="#"
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="description"
          >
            Description
            <span
              class="c5"
              title="Required"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="description_describedby"
            aria-invalid="false"
            aria-required="true"
            class="c6"
            id="description"
            name="description"
            type="text"
            value="Apples"
          />
          <div
            class="c7"
          />
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="cost"
          >
            Cost
            <span
              class="c5"
              title="Required"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="cost_describedby"
            aria-invalid="false"
            aria-required="true"
            class="c6"
            id="cost"
            name="cost"
            type="number"
            value="10"
          />
          <div
            class="c7"
          />
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="quantity"
          >
            Quantity
            <span
              class="c5"
              title="Required"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="quantity_describedby"
            aria-invalid="false"
            aria-required="true"
            class="c6"
            id="quantity"
            name="quantity"
            type="number"
            value="2"
          />
          <div
            class="c7"
          />
        </div>
        <button
          class="sc-AxjAm c8"
          type="submit"
        >
          Add item
        </button>
      </form>
      <table
        class="c9"
      >
        <thead>
          <tr>
            <th
              class="c10"
            >
              Description
            </th>
            <th
              class="c10"
            >
              Cost
            </th>
            <th
              class="c10"
            >
              Quantity
            </th>
            <th
              class="c10"
            >
              Price
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="c11"
          >
            <td
              class="c12"
            >
              Apples
            </td>
            <td
              class="c12"
            >
              $
              10
            </td>
            <td
              class="c12"
            >
              2
            </td>
            <td
              class="c12"
            >
              $
              20
            </td>
          </tr>
        </tbody>
      </table>
      <p
        class="c13"
      >
        Total: $20
      </p>
      <div
        class="c14"
      >
        <button
          aria-disabled="false"
          class="sc-AxjAm c15"
          type="button"
        >
          Submit invoice
        </button>
      </div>
    </article>
  </main>
</div>
`;

exports[`When rendered then the HTML matches this snapshot 1`] = `
.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #191919;
  font-size: 0.9rem;
}

.c6 {
  width: 100%;
  border: 1px solid #DAE5EC;
  background-color: white;
  border-radius: 0.25rem;
  line-height: 1.3em;
  font-size: 0.9rem;
  text-indent: 0.8rem;
  padding: 0;
  box-sizing: border-box;
  font-family: Roboto,Arial,Helvetica,sans-serif;
  color: #191919;
  height: 2.5rem;
}

.c7 {
  line-height: 0.9rem;
  padding-bottom: 1rem;
  padding-top: 0.2rem;
  font-size: 0.8rem;
}

.c3 {
  display: grid;
  grid-template-rows: 1.25rem minmax(2.5rem,auto) minmax(2.1rem,auto);
  box-sizing: border-box;
}

.c5 {
  margin-left: 0.25rem;
  color: #E23636;
}

.c9 {
  border-collapse: collapse;
  width: 100%;
  border: 1px solid #EBF6FF;
}

.c10 {
  padding: 0.5rem;
  border-bottom: 1px solid #DAE5EC;
  color: #191919;
  opacity: 0.75;
  font-size: 0.8rem;
  text-transform: uppercase;
  text-align: left;
}

.c2 {
  display: grid;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  grid-template-columns: repeat(auto-fit,minmax(12.5rem,1fr));
  grid-row-gap: 0.5rem;
  grid-column-gap: 0.5rem;
  margin-bottom: 2rem;
}

.c0 {
  background-color: #e9ebec;
  border-top: 0.5rem solid #00b7e2;
  width: 100%;
  font-family: Roboto,Arial,Helvetica,sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: auto;
  max-width: 50%;
  min-width: 18.75rem;
  min-height: 31.25rem;
  margin-top: 5%;
  background-color: #fafafa;
  box-shadow: 0 -2px 0.75rem rgba(0,0,0,0.03);
  padding: 1rem;
}

.c11 {
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  margin-top: 2rem;
  -webkit-align-content: end;
  -ms-flex-line-pack: end;
  align-content: end;
  display: grid;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: end;
  -webkit-justify-content: end;
  -ms-flex-pack: end;
  justify-content: end;
}

.c13 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 2.5rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all ease-in-out 0.2s;
  transition: all ease-in-out 0.2s;
  vertical-align: top;
  white-space: nowrap;
  font-weight: bold;
  padding: 0 1.5rem;
  position: relative;
  font-size: 1rem;
  text-transform: capitalize;
  -webkit-letter-spacing: 0.01rem;
  -moz-letter-spacing: 0.01rem;
  -ms-letter-spacing: 0.01rem;
  letter-spacing: 0.01rem;
  font-family: inherit;
  color: white;
  background-color: #00b7e2;
}

.c13[aria-disabled='true'] {
  opacity: 0.4;
  pointer-events: none;
}

.c13:hover {
  background-color: #0097bb;
}

.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 2.5rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all ease-in-out 0.2s;
  transition: all ease-in-out 0.2s;
  vertical-align: top;
  white-space: nowrap;
  font-weight: bold;
  padding: 0 1.5rem;
  position: relative;
  font-size: 1rem;
  text-transform: capitalize;
  -webkit-letter-spacing: 0.01rem;
  -moz-letter-spacing: 0.01rem;
  -ms-letter-spacing: 0.01rem;
  letter-spacing: 0.01rem;
  font-family: inherit;
  color: #00b7e2;
  background-color: white;
  border: 1px solid #00b7e2;
  padding: 0 0.9rem;
  -webkit-align-self: start;
  -ms-flex-item-align: start;
  align-self: start;
  margin-top: 1.25rem;
}

.c8[aria-disabled='true'] {
  opacity: 0.4;
  pointer-events: none;
}

.c8:hover {
  background-color: #00b7e2;
  color: white;
}

.c12 {
  display: grid;
  justify-items: end;
}

<div>
  <main
    class="c0"
  >
    <article
      class="c1"
    >
      <form
        action="#"
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="description"
          >
            Description
            <span
              class="c5"
              title="Required"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="description_describedby"
            aria-invalid="false"
            aria-required="true"
            class="c6"
            id="description"
            name="description"
            type="text"
            value=""
          />
          <div
            class="c7"
          />
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="cost"
          >
            Cost
            <span
              class="c5"
              title="Required"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="cost_describedby"
            aria-invalid="false"
            aria-required="true"
            class="c6"
            id="cost"
            name="cost"
            type="number"
            value="0"
          />
          <div
            class="c7"
          />
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="quantity"
          >
            Quantity
            <span
              class="c5"
              title="Required"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="quantity_describedby"
            aria-invalid="false"
            aria-required="true"
            class="c6"
            id="quantity"
            name="quantity"
            type="number"
            value="0"
          />
          <div
            class="c7"
          />
        </div>
        <button
          class="sc-AxjAm c8"
          type="submit"
        >
          Add item
        </button>
      </form>
      <table
        class="c9"
      >
        <thead>
          <tr>
            <th
              class="c10"
            >
              Description
            </th>
            <th
              class="c10"
            >
              Cost
            </th>
            <th
              class="c10"
            >
              Quantity
            </th>
            <th
              class="c10"
            >
              Price
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
      <p
        class="c11"
      >
        Total: $0
      </p>
      <div
        class="c12"
      >
        <button
          aria-disabled="true"
          class="sc-AxjAm c13"
          type="button"
        >
          Submit invoice
        </button>
      </div>
    </article>
  </main>
</div>
`;
